pilot/pkg/model/endpointshards.go

```diff
@@ -0,0 +1,216 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+
+	"istio.io/istio/pilot/pkg/serviceregistry/provider"
+	"istio.io/istio/pkg/cluster"
+	"istio.io/istio/pkg/config/schema/gvk"
+	"istio.io/istio/pkg/util/sets"
+)
+
+// shardRegistry is a simplified interface for registries that can produce a shard key
+type shardRegistry interface {
+	Cluster() cluster.ID
+	Provider() provider.ID
+}
+
+// ShardKeyFromRegistry computes the shard key based on provider type and cluster id.
+func ShardKeyFromRegistry(instance shardRegistry) ShardKey {
+	return ShardKey{Cluster: instance.Cluster(), Provider: instance.Provider()}
+}
+
+// ShardKey is the key for EndpointShards made of a key with the format "cluster/provider"
+type ShardKey struct {
+	Cluster  cluster.ID
+	Provider provider.ID
+}
+
+func (sk ShardKey) String() string {
+	return fmt.Sprintf("%s/%s", sk.Provider, sk.Cluster)
+}
+
+// EndpointShards holds the set of endpoint shards of a service. Registries update
+// individual shards incrementally. The shards are aggregated and split into
+// clusters when a push for the specific cluster is needed.
+type EndpointShards struct {
+	// mutex protecting below map.
+	sync.RWMutex
+
+	// Shards is used to track the shards. EDS updates are grouped by shard.
+	// Current implementation uses the registry name as key - in multicluster this is the
+	// name of the k8s cluster, derived from the config (secret).
+	Shards map[ShardKey][]*IstioEndpoint
+
+	// ServiceAccounts has the concatenation of all service accounts seen so far in endpoints.
+	// This is updated on push, based on shards. If the previous list is different than
+	// current list, a full push will be forced, to trigger a secure naming update.
+	// Due to the larger time, it is still possible that connection errors will occur while
+	// CDS is updated.
+	ServiceAccounts sets.Set
+}
+
+// Keys gives a sorted list of keys for EndpointShards.Shards.
+// Calls to Keys should be guarded with a lock on the EndpointShards.
+func (es *EndpointShards) Keys() []ShardKey {
+	// len(shards) ~= number of remote clusters which isn't too large, doing this sort frequently
+	// shouldn't be too problematic. If it becomes an issue we can cache it in the EndpointShards struct.
+	keys := make([]ShardKey, 0, len(es.Shards))
+	for k := range es.Shards {
+		keys = append(keys, k)
+	}
+	if len(keys) >= 2 {
+		sort.Slice(keys, func(i, j int) bool {
+			if keys[i].Provider == keys[j].Provider {
+				return keys[i].Cluster < keys[j].Cluster
+			}
+			return keys[i].Provider < keys[j].Provider
+		})
+	}
+	return keys
+}
+
+// EndpointIndex is a mutex protected index of endpoint shards
+type EndpointIndex struct {
+	mu sync.RWMutex
+	// keyed by svc then ns
+	shardsBySvc map[string]map[string]*EndpointShards
+	// We'll need to clear the cache in-sync with endpoint shards modifications.
+	cache XdsCache
```

Comment:
I am not sure why cache need in sync with endpoint shards?

I am asking because this makes it hard to maintain

Reply:
If this is coupled with xds cache, i donot think it should be moved to `model`

Reply:
A struct in model should be a simple prototype, and can be reused, if coupled with xds cache, this is absolutely not general